netlogo-gui/src/main/app/codetab/ProceduresMenu.scala

```diff
@@ -16,22 +18,81 @@ class ProceduresMenu(target: ProceduresMenuTarget)
-        item.addActionListener{() =>
-          // invokeLater for the scrolling behavior we want. we scroll twice: first bring the end into
-          // view, then bring the beginning into view, so then we can see both, if they fit - ST 11/4/04
-          target.select(end, end)
-          EventQueue.invokeLater{() =>
-            target.select(namePos, namePos + proc.size)  // highlight the name
-          }
+
+    val items = procs.map { proc =>
+      val item = new JMenuItem(proc)
+      val namePos = procsTable(proc).identifier.start
+      val end  = procsTable(proc).endKeyword.end
+      item.addActionListener{() =>
+        // invokeLater for the scrolling behavior we want. we scroll twice: first bring the end into
+        // view, then bring the beginning into view, so then we can see both, if they fit - ST 11/4/04
+        target.select(end, end)
+        EventQueue.invokeLater{() =>
+          target.select(namePos, namePos + proc.length)  // highlight the name
         }
-        menu.add(item)
       }
+      item
     }
+
+    val filterField = new JTextField
+    filterField.getDocument.addDocumentListener(new DocumentListener {
+      override def removeUpdate(e: DocumentEvent): Unit = changedUpdate(e)
+      override def insertUpdate(e: DocumentEvent): Unit = changedUpdate(e)
+      override def changedUpdate(e: DocumentEvent): Unit = {
+        repopulate(menu, filterField, items)
+        menu.getSubElements.collectFirst {
+          case it: JMenuItem => MenuSelectionManager.defaultManager.setSelectedPath(Array(menu, it))
+          case _ =>
+        }
+
+        // Changing the number of items changes the desired size of the menu
+        menu.validate() // recalculate correct size
+        menu.setSize(menu.getPreferredSize) // resize to that size
+        Option(
+          if (System.getProperty("os.name").startsWith("Mac")) SwingUtilities.getRoot(menu)
+          else menu.getParent
+        ).foreach(r => r.setSize(r.getPreferredSize)) // resize the popup to match
+        // All of the above is necessary.
+        // -BCH 1/29/2018
+      }
+    })
+
+    filterField.addKeyListener(new KeyAdapter {
+      override def keyReleased(e: KeyEvent): Unit = keyPressed(e)
+      override def keyTyped(e: KeyEvent): Unit = keyPressed(e)
+      override def keyPressed(e: KeyEvent): Unit = {
+        // TBH, it seems like you should just be able to do:
+        // MenuSelectionManager.defaultManager().processKeyEvent(e)
+        // here and it should work. Not sure why it doesn't, but meh. -BCH 1/29/2018
+        menu.dispatchEvent(e)
+        if (e.getKeyCode == KeyEvent.VK_ENTER) {
+          val path = MenuSelectionManager.defaultManager().getSelectedPath
+          if (path.nonEmpty) path.last match {
+            case it: JMenuItem if it.isArmed =>
+              it.doClick()
+              menu.setVisible(false)
+            case _ =>
+          }
+        }
+      }
+    })
```

Comment:
I would probably *prefer* to see `filterField` become a subclass of `JTextField` and see `repopulate` as well as most of the logic in L38-78 get moved to that class (which would hold references to `menu` and `items`). Having said that, I didn't write this code myself so (a) I don't know whether that is possible in some way that actually leaves the resulting system simpler, and (b) doesn't take so much time to figure out that it makes this whole issue not worth finishing.

Reply:
I tried a few variants of this (actually, subclassing the `JPopupMenu` as `FilterablePopupMenu`) but it turned out to be somewhat awkward without much improvement, so I think I might leave it as is for now. If we want to apply this to other menus in the future, we could refactor then.